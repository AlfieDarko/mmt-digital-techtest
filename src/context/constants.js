export const CLEAR_BASKET = "CLEAR_BASKET"
export const CHANGE_QUANTITY = "CHANGE_QUANTITY"