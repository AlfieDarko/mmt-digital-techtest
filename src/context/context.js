import { createContext, useContext } from 'react'

export const ProductBasketStateCtx = createContext()
export const ProductBasketDispatchCtx = createContext()


export const useProductBasketState = () => {
  const context = useContext(ProductBasketStateCtx);
  if (context === undefined) {
    throw new Error(
      "useProductBasketState() must be used within a ProductBasket Provider. Check you are using it correctly."
    );
  }
  return context;
}

export const useProductBasketDispatch = () => {
  const context = useContext(ProductBasketDispatchCtx);
  if (context === undefined) {
    throw new Error(
      "useProductBasketDispatch() must be used within a ProductBasket Provider. Check you are using it correctly."
    );
  }
  return context;
}

export const useProductBasket = () => {
  return [useProductBasketState(), useProductBasketDispatch()]
}
