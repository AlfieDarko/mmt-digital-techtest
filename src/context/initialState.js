export const initBasketState = {
  products: {
    byId: {
      "drinkid1": {
        id: 'drinkid1',
        title: "Mountain Dew",
        description: "Sugar, Citric acid with a splash of Hydrofluoric Acid & water",
        price: 109,
        quantity: 0,
        imageSRC: "https://digitalcontent.api.tesco.com/v2/media/ghs/5cc66364-c063-48f1-b9e0-8d06bcb5b857/snapshotimagehandler_1451681747.jpeg"
      },
      "drinkid2": {
        id: 'drinkid2',
        title: "Desperados",
        description: "Tequila flavoured, Mexican inspired, French created, Croatian produced beer.",
        price: 599,
        quantity: 0,
        imageSRC: "https://www.thedesignateddriver.co.uk/wp-content/uploads/2019/08/Desperados-600x600.jpg"
      },
      "drinkid3": {
        id: 'drinkid3',
        title: "Jack Daniels",
        description: "Moonshine but rebranded.",
        price: 3499,
        quantity: 0,
        imageSRC: "https://www.latramuntana.store/545-large_default/jack-daniels-amb-balanci-3-l.jpg"
      }
    },
    allIds: ['drinkid1', 'drinkid2', 'drinkid3']
  }
}