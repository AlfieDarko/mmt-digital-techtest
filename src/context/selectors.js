export const selectProductById = (state, id) => {
  return state?.products.byId[id]
}

export const getNewProductQuantity = (state, id, changeInQuantity) => {
  const product = selectProductById(state, id)
  if (product.quantity === 0 && changeInQuantity < 0) {
    throw new Error('Cannot reduce quantity below 0. Please try again')
  }
  return product.quantity + changeInQuantity
}

export const getProducts = (state) => {
  return state.products.byId
}