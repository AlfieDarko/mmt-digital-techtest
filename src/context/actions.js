import { CLEAR_BASKET, CHANGE_QUANTITY } from './constants'

export const clearBasket = (dispatch) => {
  dispatch({ type: CLEAR_BASKET })
}

export const changeProductQuantity = (id, quantity, dispatch) => {
  dispatch({ type: CHANGE_QUANTITY, payload: { id, quantity } })
}