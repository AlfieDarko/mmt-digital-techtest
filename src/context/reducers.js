import React from 'react'
import { initBasketState } from './initialState'

export const productBasketReducer = (state, action) => {

  switch (action.type) {
    case "CHANGE_QUANTITY": {
      const { payload: { id, quantity } } = action;
      return {
        ...state, products: {
          ...state.products, byId: {
            ...state.products.byId, [id]: {
              ...state.products.byId[id], quantity
            }
          }
        }
      }
    }

    case "CLEAR_BASKET": {
      return {
        ...initBasketState
      }
    }

    default: {
      return state
    }
  }
}