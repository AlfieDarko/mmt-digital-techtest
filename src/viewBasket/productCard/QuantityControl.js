import React from 'react'
import PropTypes from 'prop-types'
import { Box, Flex, Icon, Button } from '@chakra-ui/core'
import { useProductBasket } from '../../context/context'
import { changeProductQuantity } from '../../context/actions'
import { getNewProductQuantity } from '../../context/selectors'

const QuantityControl = ({ id, title, quantity }) => {
  const [state, dispatch] = useProductBasket()

  const onClickHandler = (id, changeInQuantity) => {
    const newQuantity = getNewProductQuantity(state, id, changeInQuantity)
    changeProductQuantity(id, newQuantity, dispatch)
  }

  return (
    <>
      <Flex data-testid="quantity" justifyContent="space-between">
        <Flex alignItems="center" pr={12} >Quantity: </Flex>
        <Box display="flex">
          <Button p="0" onClick={() => onClickHandler(id, +1)} data-testid={`increase-quantity-btn-${id}`}>
            <Icon name="triangle-up" size="20px" aria-label={`increase quantity of ${title}`} />
          </Button>
          <Flex alignItems="center" px="3" data-testid={`product-quantity-${id}`}>{quantity}</Flex>

          <Button p="0" disabled={quantity === 0} data-testid={`decrease-quantity-btn-${id}`} alt={`decrease quantity of ${title}`} onClick={() => onClickHandler(id, -1)} >
            <Icon name="triangle-down" size="20px" aria-label={`decrease quantity of ${title}`} />
          </Button>
        </Box>
      </Flex>
    </>
  )
}

QuantityControl.propTypes = {
  title: PropTypes.string.isRequired,
  quantity: PropTypes.number.isRequired,
  id: PropTypes.string.isRequired
}


export default QuantityControl
