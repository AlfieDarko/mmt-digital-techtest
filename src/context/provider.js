import React from 'react'
import { ProductBasketStateCtx, ProductBasketDispatchCtx } from './context'
import { productBasketReducer } from './reducers'
import { useReducer } from 'react'
import { initBasketState } from './initialState'

export const ProductBasketProvider = ({ children }) => {
  const [state, dispatch] = useReducer(productBasketReducer, initBasketState)

  return (
    <ProductBasketStateCtx.Provider value={state}>
      <ProductBasketDispatchCtx.Provider value={dispatch}>
        {children}
      </ProductBasketDispatchCtx.Provider>
    </ProductBasketStateCtx.Provider>
  )
}

