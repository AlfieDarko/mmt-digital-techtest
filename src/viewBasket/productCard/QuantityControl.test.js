import React from 'react'
import QuantityControl from './QuantityControl'
import { render } from '@testing-library/react'
import { ProductBasketProvider } from '../../context/provider'
import { ThemeProvider } from '@chakra-ui/core'

const customRender = (ui, {
  providerProps, ...renderOptions
}) => {
  return render(
    <ProductBasketProvider {...providerProps}>
      <ThemeProvider>
        {ui}
      </ThemeProvider>
    </ProductBasketProvider>, renderOptions
  )
}

describe('QuantityControl', () => {
  const props = {
    imageSRC: "",
    title: "test product",
    description: "product description",
    price: 100,
    quantity: 2,
    id: 'beer1'
  }

  const providerProps = {
    state: {
      products: {
        byId: {
          "beer1": props
        }
      }
    }
  }

  it('Should match snapshots', () => {
    const { asFragment } = customRender(<QuantityControl {...props} />, { providerProps });
    expect(asFragment()).toMatchSnapshot();
  });
});