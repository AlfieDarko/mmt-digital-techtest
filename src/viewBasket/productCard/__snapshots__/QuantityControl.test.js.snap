// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuantityControl Should match snapshots 1`] = `
<DocumentFragment>
  <div
    class="css-gg4vpm"
    data-testid="quantity"
  >
    <div
      class="css-hgou8i"
    >
      Quantity: 
    </div>
    <div
      class="css-k008qs"
    >
      <button
        class="css-1wf4v31"
        data-testid="increase-quantity-btn-beer1"
        type="button"
      >
        <svg
          aria-label="increase quantity of test product"
          class="css-zdnkem"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M12.8,5.4c-0.377-0.504-1.223-0.504-1.6,0l-9,12c-0.228,0.303-0.264,0.708-0.095,1.047 C2.275,18.786,2.621,19,3,19h18c0.379,0,0.725-0.214,0.895-0.553c0.169-0.339,0.133-0.744-0.095-1.047L12.8,5.4z"
            fill="currentColor"
          />
        </svg>
      </button>
      <div
        class="css-rht8x5"
        data-testid="product-quantity-beer1"
      >
        2
      </div>
      <button
        alt="decrease quantity of test product"
        class="css-1wf4v31"
        data-testid="decrease-quantity-btn-beer1"
        type="button"
      >
        <svg
          aria-label="decrease quantity of test product"
          class="css-zdnkem"
          focusable="false"
          role="presentation"
          viewBox="0 0 24 24"
        >
          <path
            d="M21,5H3C2.621,5,2.275,5.214,2.105,5.553C1.937,5.892,1.973,6.297,2.2,6.6l9,12 c0.188,0.252,0.485,0.4,0.8,0.4s0.611-0.148,0.8-0.4l9-12c0.228-0.303,0.264-0.708,0.095-1.047C21.725,5.214,21.379,5,21,5z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
</DocumentFragment>
`;
